src/devices/SoftwareSpi/SoftwareSpi.cs

```diff
@@ -0,0 +1,199 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Threading;
+using System.Device.Gpio;
+using System.Device.Spi;
+
+namespace Iot.Device.Spi
+{
+    public class SoftwareSpi : SpiDevice
+    {
+        private GpioController _controller;
+        private readonly int _clk;
+        private readonly int _miso;
+        private readonly int _mosi;
+        private readonly int _cs;
+        private readonly SpiConnectionSettings _settings;
+
+        public SoftwareSpi(int clk, int miso, int mosi, int cs, SpiConnectionSettings settings = null, GpioController controller = null)
+        {
+            _controller = controller ?? new GpioController();
+
+            _settings = settings ?? new SpiConnectionSettings(-1, -1);
+
+            _clk = clk;
+            _miso = miso;
+            _mosi = mosi;
+            _cs = cs;
+
+            _controller.OpenPin(_clk, PinMode.Output);
+            _controller.OpenPin(_miso, PinMode.Input);
+            _controller.OpenPin(_mosi, PinMode.Output);
+            _controller.OpenPin(_cs, PinMode.Output);
+
+            // aka. CPOL
+            bool idle = ((int)_settings.Mode & 1) == 1;
+
+            // aka. CPHA
+            bool cpha = ((int)_settings.Mode & 0b10) == 0b10;
+
+            _controller.Write(_cs,!(bool)_settings.ChipSelectLineActiveState);
+            _controller.Write(_clk, idle);
+
+            // TODO: To respect ClockFrequency we need to inject the right delays here
+            //       and have some very accurate way to measure time.
+            //       Ideally we should verify the output with an oscilloscope.
+            if (cpha)
+            {
+                _bitTransfer = new ScopeData(
```

Comment:
I think this whole section with cpha should be commented much more so that it is more readable. CPOL idle is very straight forward, but the CPHA shiftness isn't so it would help if you add verbose comments in here describing the process in case there is a bug here in the future and somebody is trying to read/modify this code.